My Store Analytics: Custom Range UI for Visitors and Conversions

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Closes #11972

For Custom Range tab, we want to redact information about Visitors and Conversions, because there is no accurate data for visits for a custom date range. However, we can show those information when an individual value in the chart is selected, since the data is accurate.


This PR also adds a new redaction view for those two labels, with an information icon. It will serve as starting point to add the bottom sheet information in task #12162 

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Start app, create a custom range tab if you haven't,
2. In the custom range, ensure that Visitors and Conversions label are redacted with an information icon added,
3. Move to the other stats tabs, ensure that Visitors and Conversions are loaded properly.
4. Go back to custom range, select a value in the range, ensure that the Visitors and Conversions data are now shown,
5. Deselect the value, ensure the values are redacted again.

## Video
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/266376/f5bb63c5-e1f6-41e2-87b6-f5978664530d



---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hafizrahman committed:
Add .redactedDueToCustomRange case for hiding visitor and conversion on custom range tab.

hafizrahman committed:
Set the custom range tab as redacted, and filter it out from the global update function.

hafizrahman committed:
Do not sync site visit stats for custom range tab.

hafizrahman committed:
Also set redacted state for conversion label.

hafizrahman committed:
Also avoid getting summary stats for custom range tab.

hafizrahman commented:
Setting back to draft temporarily to pivot this PR to add functionality

hafizrahman committed:
Revert the logic to hide visit/summary stats to custom tabs.

We actually need to do this still, just redact the value from overall range.

hafizrahman committed:
In custom range, show visitor stats when a bar is selected and redact if deselected.

hafizrahman commented:
This should be good for review now @itsmeichigo thank you!

itsmeichigo commented:
This looks good for sites with Jetpack plugin, but for sites with Jetpack Connection Package only, we should keep the redact view for Jetpack:

\ | Before | After this PR 
--- | --- | ---
No selected date | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/35a7453b-7582-485d-94aa-0810e5376f53" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/cfea642f-8d72-49b4-bd5c-8857132d24af" width=320 /> 
With selected date | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/d5018561-b237-47bd-8912-f379ae9b8030" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/fdc32162-a993-47b0-a8e0-baf60c5b7737" width=320 />

Same issue happens to self-hosted sites without Jetpack:

| Before | After this PR | 
| --- | --- |
| <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/cf37af75-3cf8-4885-a17f-53ebcb644364" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/4b275e30-00a0-4865-bd19-ed64d58ddfa7" width=320 /> |

Since we are adding a bottom sheet explaining the unavailability of the visit stats due to the arbitrary time range, we should exclude the cases for JCP and non-JP sites.

hafizrahman committed:
Add logic to handle redaction in custom range depending on site type.

Specifically:
- Jetpack CP sites should stay having the .redactedDueToJetpack state.
- Self-hosted sites with no Jetpack should have the .hidden state.
- Jetpack CP sites or self-hosted non-Jetpack sites should not update state on bar selection.

hafizrahman commented:
@itsmeichigo thank you for considering the use case for Jetpack CP and no-Jetpack sites. The screenshots you provided are also very useful. I hadn't considered "for sites with Jetpack Connection Package only, we should keep the redact view for Jetpack" before, but it makes sense!

I added the state handling logic both for tab creation and bar selection on https://github.com/woocommerce/woocommerce-ios/pull/12164/commits/aa345a83426e335e726bf3bc947c9e59a0c0ea33

hafizrahman committed:
lint fix.

hafizrahman committed:
Merge branch 'trunk'

hafizrahman commented:
Should be good for further review including JCP and non-Jetpack site cases @itsmeichigo

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -502,6 +503,23 @@ private extension StoreStatsAndTopPerformersViewController {
+        } else {
+            if site.isJetpackCPConnected {
+                customRangeVC.siteVisitStatsMode = .redactedDueToJetpack
+            }
+            else {
```
Super nit:

```suggestion
            } else {
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsDataOrRedactedView.swift:
```diff
@@ -70,6 +72,8 @@ private extension StoreStatsDataOrRedactedView {
         case .redacted, .redactedDueToJetpack:
             redactedView.showJetpackImage = state == .redactedDueToJetpack
+        case .redactedDueToCustomRange:
+            redactedView.showInfoIcon = state == .redactedDueToCustomRange
```
We should not need to compare the state here:

```suggestion
            redactedView.showInfoIcon = true
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsEmptyView.swift:
```diff
@@ -9,8 +9,17 @@ final class StoreStatsEmptyView: UIView {
     }
 
+    /// Whether to show information icon
+    var showInfoIcon: Bool = false {
```
Super nit:

```suggestion
    var showInfoIcon = false {
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsPeriodViewModel.swift:
```diff
@@ -237,6 +239,8 @@ private extension StoreStatsPeriodViewModel {
         case .redacted:
             return .redacted
+        case .redactedDueToCustomRange:
+            return .redactedDueToCustomRange
```
We might want to follow the JCP case to display the redacted view for the conversion stat and display the info icon only for the visitor stat.

hafizrahman replied:
Added in c31913d 👍🏼

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift:
```diff
@@ -485,6 +490,25 @@ private extension StoreStatsV4PeriodViewController {
+
+            guard let site = stores.sessionManager.defaultSite else { return }
+
+            if site.isWordPressComStore || (site.isJetpackConnected && !site.isJetpackCPConnected) {
+                if selectedIndex != nil && siteVisitStatsMode == .redactedDueToCustomRange {
+                    siteVisitStatsMode = .default
+                } else if selectedIndex == nil && siteVisitStatsMode == .default {
+                    siteVisitStatsMode = .redactedDueToCustomRange
+                }
+            }
```
This can be simplified for better readability:

```suggestion
            if site.isJetpackConnected && site.isJetpackThePluginInstalled {
                siteVisitStatsMode = selectedIndex != nil ? .default : .redactedDueToCustomRange
            }
```

itsmeichigo reviewed (approved):
This works quite nicely! I'd suggest we follow the JCP case to display the info icon only for the visitor stat, while leaving the redacted view for the conversion stat.

Also left some other code suggestions below, but I'm pre-approving this PR to avoid blocking development.

hafizrahman committed:
else spacing

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
remove unneeded type

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
refactor showInfoIcon conditional

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Simplify stats mode toggle logic

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Show simple redaction without icon for conversion, matching jetpack CP behavior.
